eureka-client/src/main/java/com/netflix/discovery/InstanceInfoReplicator.java

```diff
@@ -0,0 +1,114 @@
+
+    private final AtomicBoolean started;
+    private final RateLimiter rateLimiter;
+    private final int burstSize;
+    private final int allowedRatePerMinute;
+
+    InstanceInfoReplicator(DiscoveryClient discoveryClient, InstanceInfo instanceInfo, int replicationIntervalSeconds, int burstSize) {
+        this.discoveryClient = discoveryClient;
+        this.instanceInfo = instanceInfo;
+        this.scheduler = Executors.newScheduledThreadPool(1,
+                new ThreadFactoryBuilder()
+                        .setNameFormat("DiscoveryClient-InstanceInfoReplicator-%d")
+                        .setDaemon(true)
+                        .build());
+
+        this.scheduledPeriodicRef = new AtomicReference<Future>();
+
+        this.started = new AtomicBoolean(false);
+        this.rateLimiter = new RateLimiter(TimeUnit.MINUTES);
+        this.replicationIntervalSeconds = replicationIntervalSeconds;
+        this.burstSize = burstSize;
+
+        this.allowedRatePerMinute = 60 * this.burstSize / this.replicationIntervalSeconds;
+        logger.info("InstanceInfoReplicator onDemand update allowed rate per min is {}", allowedRatePerMinute);
+    }
+
+    public void start(int initialDelayMs) {
+        if (started.compareAndSet(false, true)) {
+            instanceInfo.setIsDirty();  // for initial register
+            scheduler.schedule(this, initialDelayMs, TimeUnit.SECONDS);
+        }
+    }
+
+    public void stop() {
+        scheduler.shutdown();
+        started.set(false);
+    }
+
+    public boolean onDemandUpdate() {
+        if (rateLimiter.acquire(burstSize, allowedRatePerMinute)) {
+            scheduler.submit(new Runnable() {
+                @Override
+                public void run() {
+                    logger.debug("Executing on-demand update of local InstanceInfo");
+
+                    // cancel the latest scheduled update, it will be rescheduled at the end of run()
+                    Future latestPeriodic = scheduledPeriodicRef.get();
+                    if (latestPeriodic != null && !latestPeriodic.isDone()) {
+                        latestPeriodic.cancel(false);
+                    }
+
+                    InstanceInfoReplicator.this.run();
+                }
+            });
+            return true;
+        } else {
+            logger.warn("Ignoring onDemand update due to rate limiter");
+            return false;
+        }
+    }
+
+    public void run() {
+        try {
+            discoveryClient.refreshInstanceInfo();
+
+            Long dirtyTimestamp = instanceInfo.isDirtyWithTime();
+            if (dirtyTimestamp != null) {
+                discoveryClient.register();
+                instanceInfo.unsetIsDirty(dirtyTimestamp);
+            }
+        } catch (Throwable t) {
+            logger.warn("There was a problem with the instance info replicator", t);
+        } finally {
+            Future next = scheduler.schedule(this, replicationIntervalSeconds, TimeUnit.SECONDS);
```

Comment:
Are you trying to achieve scheduling with fixed delay with this schedule-at-the-end-of-a-run construct?